modules/catalog/src/main/java/org/apache/ignite/internal/catalog/ClockWaiter.java

```diff
@@ -125,47 +152,45 @@ public CompletableFuture<Void> waitFor(HybridTimestamp targetTimestamp) {
     }
 
     private CompletableFuture<Void> doWaitFor(HybridTimestamp targetTimestamp) {
-        CompletableFuture<Void> future = nowTracker.waitFor(targetTimestamp.longValue());
+        HybridTimestamp now = clock.now();
 
-        ScheduledFuture<?> scheduledFuture;
+        if (targetTimestamp.compareTo(now) <= 0) {
+            return completedFuture(null);
+        }
 
-        if (!future.isDone()) {
-            // This triggers a clock update.
-            HybridTimestamp now = clock.now();
+        CompletableFuture<Void> future = nowTracker.waitFor(targetTimestamp.longValue());
 
-            if (targetTimestamp.compareTo(now) <= 0) {
-                assert future.isDone();
+        // Adding 1 to account for a possible non-null logical part of the targetTimestamp.
+        long millisToWait = targetTimestamp.getPhysical() - now.getPhysical() + 1;
 
-                scheduledFuture = null;
-            } else {
-                // Adding 1 to account for a possible non-null logical part of the targetTimestamp.
-                long millisToWait = targetTimestamp.getPhysical() - now.getPhysical() + 1;
+        ScheduledFuture<?> scheduledFuture = scheduler.schedule(triggerClockUpdate, millisToWait, TimeUnit.MILLISECONDS);
 
-                scheduledFuture = scheduler.schedule(this::triggerClockUpdate, millisToWait, TimeUnit.MILLISECONDS);
-            }
-        } else {
-            scheduledFuture = null;
-        }
+        return future
+                .handle((res, ex) -> {
+                    scheduledFuture.cancel(true);
 
-        return future.handle((res, ex) -> {
-            if (scheduledFuture != null) {
-                scheduledFuture.cancel(true);
-            }
+                    translateTrackerClosedException(ex);
```

Comment:
This looks like we are handling the error every time, but it might not exist. Maybe pull `if (ex != null)` out of the method?